packages/react-compose/src/createOptionsResolver.ts

```diff
@@ -0,0 +1,67 @@
+import { ComposePreparedOptions, GenericDictionary } from './types';
+import { getNativeElementProps, assign } from '@uifabric/utilities';
+import { resolveClasses } from './resolveClasses';
+import { resolveSlotProps } from './resolveSlotProps';
+
+export type OptionsResolverResult = {
+  state: GenericDictionary;
+  slots: GenericDictionary;
+  slotProps: GenericDictionary;
+};
+
+/**
+ * A set of mapped props for intrinsic element types.
+ */
+export const defaultMappedProps: Record<string, string> = {
+  iframe: 'src',
+  img: 'src',
+  input: 'type',
+};
+
+export const EmptyRender = () => null;
+
+/**
+ * Creates an options resolve function which should attach to an options object
+ * for a composed component.
+ */
+export const createOptionsResolver = <TState>(options: ComposePreparedOptions) => {
+  // Returning a function so that "resolve" function shows up in profiling.
+  // tslint:disable-next-line:no-function-expression
+  return function resolve(stateOrProps: TState): OptionsResolverResult {
+    const slotProps: OptionsResolverResult['slotProps'] = {};
+    const state = stateOrProps as OptionsResolverResult['state'];
+    const slots: OptionsResolverResult['slots'] = {
+      ...options.slots,
+      root: state.as || options.slots.root || 'div',
+    };
+    const result: OptionsResolverResult = {
+      state,
+      slotProps,
+      slots,
+    };
+
+    // Mix unrecognized props onto root, excluding the handled props.
+    assignToMapObject(
+      slotProps,
+      'root',
+      getNativeElementProps(state.as, state, [...options.handledProps, 'className']),
+    );
+
+    // Resolve slotProps/slots from state.
+    resolveSlotProps(result);
+
+    // Resolve classes.
+    resolveClasses(options.classes, result);
```

Comment:
a bit nervous about perf - i don't see much opportunity to memoize

Reply:
the logic in this function isn't substantially more expensive than manually cx'ing all things together. it loops through classes, sticks them in arrays, joins the arrays.

I think caching isn't going to help here

Reply:
@dzearing it does seem like this adds another loop through all of the slots and all of the classes. Additionally that loop will grow every-time a component is recomposed since it will have to go through all of the old classes and all of the new classes right?

Reply:
yes, but this is not a concern. think of it like this:

In sass components like legacy Calendar, we'd have this:


```tsx
render() {
return (
  <div className={ css(styles.root, props.primary && styles.primary) }>
    <div className={ css(styles.slot1, etc) }
```

This is exactly the same performance as something like this:

```tsx
const deriveClasses = (state) => ({
  root: css(styles.root, state.primary && styles.primary),
  slot1: css(styles.slot1),
});

render() {
  const classes = deriveClasses(state);

  return (
    <div className={classes.root}>
       <div className={classes.slot1}>
 ```

In a recompose scenario, we may have multiple of these deriving functions. We need to execute each of them and join them up. There is no reducing this at the high level, but that's irrelevant because the overhead of 3 functions over 2 functions is nearly 0 cost. It's what's *in the function* that's costly.

So in css in js world, we achieve fast perf, not by optimizing it at this layer, but within the class resolving process:

```tsx
const deriveClasses = state => {
  const theme = useTheme();

  // Call the function and mergeStyleSets only if the keys within the style props (2nd arg) have mutated.
  return memoizedStyles(styleProps) => {}, {theme, state.primary});
};
```

We can now call a number of these functions. The top level logic will append classnames together. 

(Small note: there is a feature in mergeStyles where we can auto expand previous registered classnames to new rules, avoiding specificity issues. We could take advantage of this by passing along previous class names to the function and including those in the memoization keys, but again, potentially hazardous if we get into some of the edge cases.)